### DISCUSSION TITLE
Is there/could there be an "official" `Model._modifications(pars)` implementation?

### DISCUSSION THREAD

**Discussion by lhenkelm** [scikit-hep/pyhf#1414] [Ideas] [answered]
`Model._modifications` ([here](https://github.com/scikit-hep/pyhf/blob/671ed56483bc602876f611cf10ff9add7f4a2137/src/pyhf/pdf.py#L463)) is an undocumented method that seems to be intended to return modifier effects given parameter values. 

This sounds very useful to me, e.g. with a use case like in [get_mc_counts in this example](https://pyhf.readthedocs.io/en/v0.6.1/examples/notebooks/binderexample/StatisticalAnalysis.html#Read-in-the-Model-from-XML-and-ROOT). If generalized, this would make it possible to do generic pre-fit and post-fit plots of distributions, purely from the data in a workspace or model specification. The only data necessary to inject from the outside would be info about the bin edges. 

However, when I tried using `Model._modifications` like that, I ran into some issues.
Apart from the leading underscore being scary, it seems it does not work on all model specs:
```python
>>> import pyhf
>>> assert pyhf.__version__ == '0.6.1'
>>> pdf = pyhf.simplemodels.hepdata_like(
...    signal_data=[12.0, 11.0],
...    bkg_data=[50.0, 52.0],
...    bkg_uncerts=[3.0, 7.0],
... )
>>> pdf._modifications(pdf.config.suggested_init())
TypeError                                 Traceback (most recent call last)
...
TypeError: only integer scalar arrays can be converted to a scalar index
>>> # it really is due to shapesys.py:168
>>> pdf.spec['channels'][0]['samples'][1]['modifiers'].pop()
{'name': 'uncorr_bkguncrt', 'type': 'shapesys', 'data': [3.0, 7.0]}
>>> pdf = pyhf.Model(pdf.spec)
>>> pdf._modifications(pdf.config.suggested_init())
([],
 [array([[[[1., 1.]],
  
          [[1., 1.]]]])])
```

So the question is: Is there already a better way to do this?
And if not, the request would be to fix `_modifications` and document it/make it official.

**Comment by alexander-held** [comment, marked as answer]
Hi, if you are looking for pre- and post-fit distributions, the `Model.main_model.expected_data` API is suitable (I think this is not documented in the API since `_MainModel` is not public?): https://github.com/scikit-hep/pyhf/blob/671ed56483bc602876f611cf10ff9add7f4a2137/src/pyhf/pdf.py#L479-L505

You might also be interested in [`cabinetry`](https://github.com/alexander-held/cabinetry/), which provides an API for pre- and post-fit plots (with uncertainties), using only a model and the data you want to overlay: [`cabinetry.visualize.data_MC`](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.visualize.data_MC) (bin edge information can be provided additionally).

**Reply by lhenkelm** [reply]
Hi, Thanks for the quick reply! I did originally have a look at `Model.expected_data` but I did not find the `return_by_sample` kwarg, which is why I turned to `_modifications`. Do I need to install from github master to access this feature? (Only tried it in the pypi releases 0.5.0 and 0.6.1)

Also thanks for the link to cabinetry! I only had time for a superficial look, but I now have two questions regarding visualize.data_MC: 
* is there an option to specify parameters not through fit results? (as in, "This is what the prediction would look like if energy resolution were at that other value" etc.) Or is it safe to hack  this in through mock fit results (providing nonsense correlation/likelihood/etc. data)?
* is there a way to get just the matplotlib axes and figure the function would save, without actually saving?

**Reply by alexander-held** [reply]
`return_by_sample` was added in #731, it should be available in versions `v0.5.2` and above (I am using it in `v0.6.1` without issues). See also #1031, and more generally #726 which is closely related to this discussion.

If you would like to know how the model looks like for some arbitrary parameter setting, then indeed you can just create a corresponding fake fit result by hand. The parameter values, uncertainties, and correlations are all needed since they all affect the plot (the latter two only changing the uncertainty band).

At the moment there is no way to get the axes/figure, but this would be great to have and is something that is planned for the next minor version (see [the `v.0.3` project](https://github.com/alexander-held/cabinetry/projects/3) for related issues). I am not an expert with `matplotlib`, so it mostly comes down to learning how to handle this properly. Any contributions there are also welcome!

Looking at the code again, some of the `cabinetry.visualize.data_MC` code could still be factored out into the `cabinetry.model_utils` module. The uncertainty calculation is already factored out there, and the calculation of the nominal values is much less complicated, but might still be nice to also factor out for people who want to use just that part without making a plot.

**Reply by alexander-held** [reply]
Reading #1031 again, I found what is likely the issue: `Model.expected_data` does not have the `return_by_sample` kwarg, only `Model.main_model.expected_data` does. @kratsg provides the reason in https://github.com/scikit-hep/pyhf/issues/1031#issuecomment-674167050.

**Reply by lhenkelm** [reply]
Yep, `Model.main_model.expected_data` is the way (from a how-to-use pyhf perspective). Thank you so much for digging that up!

**Reply by lhenkelm** [reply]
PS: Regarding the cabinetry comments,  I am no kind of expert. As a user/outsider, I really like the [seaborn](http://seaborn.pydata.org/) approach, where the API usually returns a matplotlib `Axes` or `Figure` object, or accepts one as an input for the data to be placed into. IMO these are the benefits:
  * As a library user I am not giving up any flexibilty by using the library: no matter how unusual, I can write any extensions/adaptions I want in the calling code, using the full power of the language.
  * In turn the library author can focus on the core business they consider most important for that plot, and don't have to anticipate quite so many concerns and requests for bells and whistles. 
  * Additionally,  writing automated tests for this kind of API is easy: you can assert on the state of the plotted-into axis object. 

of course, this is just, like, my opinion and I am clueless about cabinetry. So there may be good reasons not to use a seaborn-like design.

**Reply by alexander-held** [reply]
I agree with all those benefits, and definitely would like to adopt something like this. The challenge for me is understanding enough about the details to get them right (the amount of discussions in #1377 shows that there is a lot that can be considered).

**Reply by lukasheinrich** [reply]
we had a pretty long discussion on thiss here

https://github.com/scikit-hep/pyhf/pull/1377

wherer we make it possible to pass custom axes, and return Artist objects so you can channge the styling after the fact

**Reply by matthewfeickert** [reply]
> where the API usually returns a matplotlib Axes or Figure object,

I think it is best in general to stick to the `matplotlib` API decisions of returning artists and not axes. I like `seaborn` and what it does, but the basic object that its API returns and operates on is a much higher level object: [`seaborn.FacetGrid`](https://seaborn.pydata.org/generated/seaborn.FacetGrid.html#seaborn.FacetGrid)

I think it is much better to not try to reinvent another plotting API. If you want a `seaborn` like API experience then take `seaborn` as a dependency and then use that as your visualization library that you build on for plotting functions. However, this will probably work poorly for most of our goals in HEP visualization and a better approach is to stick to the `matplotlib` API.

**Reply by lhenkelm** [reply]
> I think it is best in general to stick to the `matplotlib` API decisions of returning artists and not axes. 

With great respect, I disagree. Unless you are trying to extend matplotlib (as in, make things matplotlib cannot be used to plot plottable),
you are not trying to solve the kind of problem mpl is trying to solve. They solve the problem of "make any plot".
A HEP plotting library is not trying to make any kind of plot you cannot make using mpl (as far as I am aware, at least).
Rather such a library is trying to provide useful "abbreviations" in the "language" already defined by matplotlib.
So a higher level seems appropriate.

I (it seems you disagree, and that is of course OK) do not think Artists are the best abstraction to phrase the abbreviations in.
People using matplotlib use it in terms of axes, figures, and pyplot (bah). 
So any student or researcher who has ever done any kind of "Intro to scientific computing in python" will have seen axes and figures, 
and if they only have seen pyplot, it takes at most 5 minutes to google for the OO equivalent. 
So the simplest thing would be to just provide a few functions, that either produce or modify axes. 

[axes have 62k hits on SO](https://www.google.com/search?q=matplotlib.axes.Axes+site:stackoverflow.com)
[artists have 4k hits on SO](https://www.google.com/search?q=matplotlib.artist.Artist+site:stackoverflow.com)

**Reply by matthewfeickert** [reply]
Yeah, these choices are hard in general. I'm also not a visualization expert, so I'm more parroting back what I've learned/heard from people who are. I think that we're also talking about different things, though. 

> A HEP plotting library 

a visualization module in any of the libraries that are focused on statistical analysis should not be viewed as plotting libraries. Ideally these will all be factored out in the future, just as `pyhf` hopes that the `pyhf.infer` module eventually will be as well. They should be focused on integrating with and using the existing visualization libraries that are being developed, and that requires that there is a consistent API that can be used between libraries, or at least a clearly understood hierarchy.

These discussions have been ongoing for quite some time already between the groups in Scikit-HEP and IRIS-HEP that are already working on visualization. Since you have differing views from what seems to be converging on I would encourage you to both check out the discussions and work on things like [`mplhep`](https://github.com/scikit-hep/mplhep) and [`hist`](https://github.com/scikit-hep/hist), to see why these decisions are being made (with a lot of help and advice from people who have actually built full visualization libraries before). But I would also encourage you to make your views heard in the GitHub Discussions there as well. 

You're views are obviously shared by many and are also arrived at through lots of time and work. They're not wrong. But I think it helps to make sure that the scope and use cases are fully understood by everyone when these discussions are being had, as otherwise it is way too easy to be talking past each other on the internet when people have different idea.

So I hope that we see a lot more discussions with you coming up in the future with @henryiii and @andrzejnovak involved as well. I think it would be great to have you in the mix. :)

**Reply by lukasheinrich** [reply]
Hi all,

i'm not sure whether there actually is any disagreementt is.. I think we all agree that it'd be great to have a HEP-specific viz library, much like `seaborn` is a viz library somewhat specialized for stats.

Whetther the API is 

```
ax = mplhep.some_plot(...) #seabornstyle
```

(what I find lacking here is that its harder to navigate back to the artists if e.g. you want to change faceoclors)

```
artists_tuple = mplhep.some_plot(...)  #mplstyle
```

(here MPL is a bit lacking in that it's harder to navigate to the `axes` 

```
container = mplhep.some_plot(...)  #mplstyle
container.axes
container.some_artist.set_fillcolor('g')
```
(I prefer this,  I still maintain having ax `axes` attr is useful )

or something else is prob. a detail (the current version inn pyhf just requires passing an `ax` argument so we circumvent this, while in the new PR we either use a passed `ax` or use `plt.gca()` (we do seem to all share a distastet of `plt` ;) ). 

As @matthewfeickert said  hopefully we will factor this out to `mplehep` which I hope becomes basically `seaborn` for HEP
I agree with @lhenkelm original point that it'd be great if cabinetry would automatically save figures but allow modification first.
